### ISSUE TITLE
fix: 140+ Neural Network ModelFamilyTests failures across 16 models

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1036]
## Summary

Running `ModelFamilyTests.NeuralNetworks` reveals **140 failures across 16 models** (181 passed, 56.4% pass rate). These are mathematical invariant tests that verify training convergence, gradient flow, output stability, and architecture correctness.

**Run command:** `dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --framework net10.0 -c Release --filter "FullyQualifiedName~ModelFamilyTests.NeuralNetworks"`

### Category 1: Test Configuration — Wrong InputShape/OutputShape (fix test files)

Models whose test files use the wrong `InputShape`/`OutputShape`, causing output dimension mismatches or loss function errors. **Fix the test file**, not the model.

| Model | Issue | Fix Needed |
|-------|-------|------------|
| **BGE** | Output is 768-dim embedding, test expects `[1,1]` | Set `OutputShape => [768]` |
| **ColBERT** | Same — embedding output mismatch | Set correct `OutputShape` |
| **TransformerEmbeddingNetwork** | Output 768 vs expected 1 | Set `OutputShape => [768]` |
| **CapsuleNetwork** | Output 784 vs expected 10 | Set `OutputShape => [784]` |
| **VoxelCNN** | Output 128 vs expected 1 | Set `OutputShape` to match 3D conv output |
| **AudioVisualCorrespondenceNetwork** | Output 512 vs expected 1 | Set `OutputShape => [512]` |
| **DenseNetNetwork** | Default `[1,4]` shape — needs `[1,3,32,32]` for conv | Set `InputShape => [1, 3, 32, 32]` |
| **MobileNetV3Network** | Same — conv layer needs 3D+ input | Set `InputShape => [1, 3, 224, 224]` |

**Files:** `tests/AiDotNet.Tests/ModelFamilyTests/NeuralNetworks/{Model}Tests.cs`

### Category 2: Incompatible Loss Function (fix model's DefaultLossFunction)

Models that use a loss function incompatible with the standard `Train(input, target)` signature.

| Model | Issue | Fix |
|-------|-------|-----|
| **AudioVisualCorrespondenceNetwork** | `ContrastiveLoss` requires paired inputs — throws on standard Train | Override `DefaultLossFunction` to return MSE or implement paired-input Train |
| **AudioVisualEventLocalizationNetwork** | Same ContrastiveLoss issue | Same fix |

**Files:** `src/NeuralNetworks/AudioVisualCorrespondenceNetwork.cs`, `src/NeuralNetworks/AudioVisualEventLocalizationNetwork.cs`

### Category 3: Real Code Bugs — Training Divergence

Models where the optimizer diverges (loss increases with more training).

| Model | Initial Loss | Final Loss | Issue |
|-------|-------------|-----------|-------|
| **GRUNeuralNetwork** | 0.104 | 0.398 | Loss increases during training |
| **RecurrentNeuralNetwork** | 0.000245 | 0.178 | Diverges after initial convergence |
| **GraphNeuralNetwork** | 232 | 1,706,959 | Massive divergence — likely learning rate or gradient explosion |
| **Autoencoder** | 0.062 | 0.067 | Slight divergence — optimizer overshooting |

**Root causes to investigate:**
- Learning rate too high for default Adam optimizer
- Missing gradient clipping
- Numerical instability in backward pass

**Files:** `src/NeuralNetworks/RecurrentNeuralNetwork.cs`, `src/NeuralNetworks/GRUNeuralNetwork.cs`, `src/NeuralNetworks/GraphNeuralNetwork.cs`, `src/NeuralNetworks/Autoencoder.cs`

### Category 4: Real Code Bugs — Collapsed Output / Dead Network

Models that produce identical output regardless of input.

| Model | Issue |
|-------|-------|
| **HTMNetwork** | Identical output for [0.1,...] and [0.9,...] — forward pass ignores input values |
| **HTMNetwork** | `TemporalMemoryLayer` shape mismatch: `[2048, 32]` vs `[2048, 1]` |

**Files:** `src/NeuralNetworks/HTMNetwork.cs`, `src/NeuralNetworks/Layers/TemporalMemoryLayer.cs:370`

### Category 5: Missing Deserialization Support

Models whose layers can't be deserialized for `Clone_ShouldProduceIdenticalOutput`.

| Model | Missing Layer Deserializer |
|-------|---------------------------|
| **CapsuleNetwork** | `PrimaryCapsuleLayer` — no known constructor in `DeserializationHelper` |
| **VoxelCNN** | `Conv3DLayer` — no known constructor |
| **GraphAttentionNetwork** | `GraphAttentionLayer` — missing constructor signature |

**File:** `src/Helpers/DeserializationHelper.cs`

### Category 6: Specialized Models Needing Custom Test Setup

Models that require special initialization beyond `new Model<double>()`.

| Model | Issue | Fix |
|-------|-------|-----|
| **ConditionalGAN** | 0 parameters, shape mismatch `[110]` vs `[4]` | Test needs proper `latentDim`, `numClasses` config |
| **MeshCNN** | `SetEdgeAdjacency` must be called before Forward | Test needs mesh setup in `CreateNetwork()` |
| **NEAT** | `Parameters_ShouldBeNonEmpty` — 0 params, needs evolution first | Override `Parameters_ShouldBeNonEmpty` or evolve before testing |

**Files:** `tests/AiDotNet.Tests/ModelFamilyTests/NeuralNetworks/ConditionalGANTests.cs`, `MeshCNNTests.cs`, `NEATTests.cs`

### Category 7: Backward Pass Bugs

Models where the backward pass crashes or produces wrong-rank gradients.

| Model | Issue | Source |
|-------|-------|--------|
| **GraphAttentionNetwork** | "Input and output gradient tensors must have the same rank" | `LayerBase.cs:2058` |

**File:** `src/NeuralNetworks/Layers/LayerBase.cs`, graph attention layer backward

## How to Fix

1. **Categories 1 & 6:** Edit the test files in `tests/AiDotNet.Tests/ModelFamilyTests/NeuralNetworks/` — set correct `InputShape`, `OutputShape`, and add model-specific setup
2. **Categories 2, 3, 4, 7:** Fix the source code in `src/NeuralNetworks/` — these are real bugs
3. **Category 5:** Register missing layer types in `src/Helpers/DeserializationHelper.cs`

**IMPORTANT:** Never weaken tests to match buggy code. Always fix the code first. See `.claude/CLAUDE.md` for coding standards.

## Test Commands

```bash
# Run all NN model family tests
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --framework net10.0 -c Release --filter "FullyQualifiedName~ModelFamilyTests.NeuralNetworks"

# Run specific model
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --framework net10.0 -c Release --filter "FullyQualifiedName~ModelFamilyTests.NeuralNetworks.RecurrentNeuralNetworkTests"
```
